## Linear Programming
- Linear programming(LP) is a method to find the maximum or the minimum in a linear mathematical model
- Each LP problem consists of the following components: an objective function, decision variables (or control variables), and constraints.

## Example: Giapetto Problem
- Giapetto, Inc. manufactures two types of furniture: chairs and tables. The manufacturer wants to maximize their weekly profit.
- \$20 of profit per chair.
- \$30 of profit per table.
- A chair requires 1 hour of finishing labor and 2 hours of carpentry labor.
- A table requires 2 hours of finishing labor and 1 hour of carpentry labor.
- Each week, Giapetto has onlu 100 finishing hours and 100 carpentry hours available.

$x1:$ number of chairs produced each week <br>
$x2:$ number of tables produced each week <br>
<br>
$max:  z=20x1 + 30x2$ (Objective Function)<br>
$s.t.  x1+2x2 \leq 100$  (Finishing Hours)<br>
$ 2x1+x2 \leq 100 $ (Carpentry Hours)<br>
$ x1 \geq 0 $ (Sign Restriction)<br>
$ x2 \geq 0 $ (Sing Restriction)

- PuLP uses LP Solvers(e.g., GLPK, COIN CLP/CBX, CPLEX, and GUROBI) to solve linear problems.

In [1]:
!pip install pulp

In [2]:
from pulp import *

In [3]:
problem = LpProblem('Giapetto', LpMaximize)  # Create a LP maximization problem
x1 = LpVariable('x1', lowBound=0)  # Create a variable x1 >= 0
x2 = LpVariable('x2', lowBound=0)  # Create another variable x2 >= 0
problem += 20*x1 + 30*x2  # Objective Function
problem += 1*x1 + 2*x2 <= 100  # Finishing hours
problem += 2*x1 + 1*x2 <= 100  # Carpentry hours
problem

Giapetto:
MAXIMIZE
20*x1 + 30*x2 + 0
SUBJECT TO
_C1: x1 + 2 x2 <= 100

_C2: 2 x1 + x2 <= 100

VARIABLES
x1 Continuous
x2 Continuous

In [4]:
status = problem.solve()  # Solve with the default solver
LpStatus[status]

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/egemeneroglu/opt/anaconda3/envs/datascience/lib/python3.8/site-packages/pulp/apis/../solverdir/cbc/osx/64/cbc /var/folders/p1/99fj6n2j4ls6wzzjfk9lzvpc0000gn/T/2cc84e662f4c4ccba9ad7ca55d62d8ba-pulp.mps max timeMode elapsed branch printingOptions all solution /var/folders/p1/99fj6n2j4ls6wzzjfk9lzvpc0000gn/T/2cc84e662f4c4ccba9ad7ca55d62d8ba-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 7 COLUMNS
At line 14 RHS
At line 17 BOUNDS
At line 18 ENDATA
Problem MODEL has 2 rows, 2 columns and 4 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Presolve 2 (0) rows, 2 (0) columns and 4 (0) elements
0  Obj -0 Dual inf 50 (2)
0  Obj -0 Dual inf 50 (2)
2  Obj 1666.6667
Optimal - objective value 1666.6667
Optimal objective 1666.666667 - 2 iterations time 0.002
Option for printingOptions changed from normal to all
Total time (CPU 

'Optimal'

In [5]:
value(x1), value(x2), value(problem.objective)  # Show the solution

(33.333333, 33.333333, 1666.6666500000001)

## Integer Value Solution

In [6]:
prob = LpProblem('Giapetto', LpMaximize)
x1 = LpVariable('x1', lowBound=0, cat='Integer')  # Integer variable x1 >= 0
x2 = LpVariable('x2', lowBound=0, cat='Integer')  # Integer variable x2 >= 0
prob += 20*x1 + 30*x2
prob += 1*x1 + 2*x2 <= 100
prob += 2*x1 + 1*x2 <= 100
prob

Giapetto:
MAXIMIZE
20*x1 + 30*x2 + 0
SUBJECT TO
_C1: x1 + 2 x2 <= 100

_C2: 2 x1 + x2 <= 100

VARIABLES
0 <= x1 Integer
0 <= x2 Integer

In [10]:
status = prob.solve()
LpStatus[status]

Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/egemeneroglu/opt/anaconda3/envs/datascience/lib/python3.8/site-packages/pulp/apis/../solverdir/cbc/osx/64/cbc /var/folders/p1/99fj6n2j4ls6wzzjfk9lzvpc0000gn/T/da5b78216640415587bc22a6c12b34db-pulp.mps max timeMode elapsed branch printingOptions all solution /var/folders/p1/99fj6n2j4ls6wzzjfk9lzvpc0000gn/T/da5b78216640415587bc22a6c12b34db-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 7 COLUMNS
At line 18 RHS
At line 21 BOUNDS
At line 24 ENDATA
Problem MODEL has 2 rows, 2 columns and 4 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Continuous objective value is 1666.67 - 0.00 seconds
Cgl0004I processed model has 2 rows, 2 columns (2 integer (0 of which binary)) and 4 elements
Cutoff increment increased from 1e-05 to 9.9999
Cbc0012I Integer solution of -1650 found by DiveCoefficient after 0 iterations and 0 nodes

'Optimal'

In [11]:
value(x1), value(x2), value(prob.objective)

(32.0, 34.0, 1660.0)

- Note the Integer Linear Programming solution is not just the rounded solution of the continuous LP solution (33, 33, 1650).